File: Data Structures & Algorithms/island-perimeter/submission-3.cpp
class Solution {
private:
    vector<vector<int>> grid;
    vector<vector<bool>> visited;
    int rows, cols;

    int dfs(int i, int j) {
        if (i < 0 || i >= rows || j < 0 || j >= cols || grid[i][j] == 0) return 1;
        if (visited[i][j]) return 0;

        visited[i][j] = true;
        return dfs(i + 1, j) + dfs(i - 1, j) + dfs(i, j + 1) + dfs(i, j - 1);
    }
public:
    int islandPerimeter(vector<vector<int>>& grid) {
        this->grid =  grid;
        rows = grid.size(); 
        cols = grid[0].size();
        visited = vector<vector<bool>>(rows, vector<bool>(cols, false));

        for (int r = 0; r < rows; ++r) {
            for (int c = 0; c < cols; ++c) {
                if (grid[r][c] == 1) {
                    return dfs(r, c);
                }
            }
        }
        return 0;
    }
};

File: Data Structures & Algorithms/rotting-fruit/submission-3.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        rows, cols = len(grid), len(grid[0])
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        visited = set()
        q = deque()

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 2:
                    q.append((r, c))
                    visited.add((r, c))

        def addFruit(r, c):
            if r < 0 or r == rows or c < 0 or c == cols or grid[r][c] != 1 or (r, c) in visited:
                return False
            q.append((r, c))
            visited.add((r, c))
            return True

        minutes = 0
        while q:
            added = False
            for i in range(len(q)):
                r, c = q.popleft()
                grid[r][c] = 2
                
                for dr, dc in directions:
                    if addFruit(r + dr, c + dc):
                        added = True
                
            minutes += added
        
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 1:
                    return -1

        return minutes

